chainer/_backprop_utils.py

```diff
@@ -1,40 +1,112 @@
+import six
+
 import chainer
 
 
-def concat_variable(gx, g_input):
-    """concatenate the inputs to a tuple of variable
+def _reduce(grad_list):
+    if not grad_list:
+        return None
+    if len(grad_list) >= 2:
+        grad_list[:] = [chainer.functions.add(*grad_list)]
+    return grad_list[0]
+
+
+def _pure(grad):
+    return [] if grad is None else [grad]
+
+
+class GradTable(object):
+
+    """Dict of nodes to references of gradients
 
-    Inputs:
-        None
-        ~chainer.Variable
-        tuple of variable
+    The gradients are stored as references to them in the backprop process. The
+    current implementation uses lists. Keep the lengths of lists <= 1 for the
+    strict accumulation of gradients. Leave them to accumulate gradients
+    lazily.
+
+    Args:
+        load_if_new (bool): read ``grad_var`` of node when the node has not
+            been added.
 
-    Outputs:
-        None: When both of gx and g_input is None
-        Variable: When one is None, and the other is variable
-        tuple of variable: Otherwise
     """
 
-    sum_gx = ()
-    if isinstance(gx, tuple):
-        sum_gx = gx
-    elif gx is not None:
-        sum_gx = gx,
+    def __init__(self, load_if_new=False):
+        self.grads = {}
+        self._load_if_new = load_if_new
+
+    def __setitem__(self, node, grad):
+        assert node is not None
+        self.grads[node] = _pure(grad)
+
+    def get_as_list(self, node):
+        assert node is not None
+        grads = self.grads
+        if node not in grads:
+            if self._load_if_new and node.creator_node is None:
+                node._check_old_style_gradient()
+                # accumulate the gradient only if the node is a leaf
+                grads[node] = _pure(node.grad_var)
+            else:
+                grads[node] = []
+        return grads[node]
+
+    def pop(self, node):
+        if node is None:
+            return None
+        grads = self.grads
+        if node in grads:
+            return _reduce(grads.pop(node))
+        if self._load_if_new:
+            return node.grad_var
+        else:
+            return None
+
+    def assert_no_grads(self):
+        for gx in self.grads.values():
+            assert gx == []
+
 
-    if isinstance(g_input, tuple):
-        sum_gx += g_input
-    elif g_input is not None:
-        sum_gx += g_input,
+def backprop_step(
+        func, target_input_indexes, grad_outputs, grad_inputs):
+    assert isinstance(target_input_indexes, tuple)
+    assert target_input_indexes == tuple(sorted(target_input_indexes))
+    assert isinstance(grad_outputs, tuple)
+    if func.backward_accumulate.__code__ \
+            is not chainer.FunctionNode.backward_accumulate.__code__:
```

Comment:
I feel this dispatching is a bit tricky and makes it hard to customize the behavior (e.g. writing `backward_accumulate` that behaves almost same as the default one with slight differences). I also think the default implementation should be written to the default `backward_accumulate` code (instead of `raise NotImplementedError`; it makes the readers hard to follow the logic when they only override `backward`). How about the following fix:

1. Allow `backward_accumulate` return a tuple of tuples instead of a tuple of variables. Each tuple contains the given grad input and the newly computed gradient that will be accumulated by the backprop logic. It allows `backward_accumulate` to choose lazy accumulation if available.
2. Pop only one variable in the list to pass to `backward_accumulate` instead of reducing them everytime. It does not hurt the performance even when the function supports fused accumulation.

The default mode may be worth optimizing in the current way, though (skip preparing `grad_inputs_tuple`, reduce the number of function calls). Anyway, I feel it's better to allow users to emulate the same behavior as the default one.

Reply:
We discussed offline and found that the code is anyway customizable by patching some methods (including this `backprop_step`). We assume that there are few users that want to customize this point and take the design simplicity and performance gain here by dispatching in an ad-hoc way.